File: packages/core/src/lock.ts
import { mkdir, open, readFile, rm } from "node:fs/promises";
import { dirname } from "node:path";
import { LocalConfigError } from "./errors.js";

export async function withRepositoryLock<T>(lockPath: string, operation: () => Promise<T>): Promise<T> {
  await mkdir(dirname(lockPath), { recursive: true });
  let handle;
  try {
    handle = await open(lockPath, "wx", 0o600);
    await handle.writeFile(JSON.stringify({ pid: process.pid, createdAt: new Date().toISOString() }));
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "EEXIST") {
      const owner = await readFile(lockPath, "utf8").catch(() => "unknown");
      throw new LocalConfigError("repository_locked", "Repository is currently being synchronized", { lockPath, owner });
    }
    throw error;
  }
  try {
    return await operation();
  } finally {
    await handle.close();
    await rm(lockPath, { force: true });
  }
}

File: internal/core/lock.go
package core

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"time"
)

func WithRepositoryLock(lockPath string, operation func() error) error {
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o700); err != nil {
		return err
	}
	file, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o600)
	if errors.Is(err, os.ErrExist) {
		owner, readErr := os.ReadFile(lockPath)
		if readErr != nil {
			owner = []byte("unknown")
		}
		return NewError(ErrRepositoryLocked, "Repository is currently being synchronized", map[string]any{"lockPath": lockPath, "owner": string(owner)})
	}
	if err != nil {
		return err
	}
	owner, _ := json.Marshal(map[string]any{"pid": os.Getpid(), "createdAt": time.Now().UTC().Format(time.RFC3339Nano)})
	if _, err := file.Write(owner); err != nil {
		_ = file.Close()
		_ = os.Remove(lockPath)
		return err
	}
	defer func() { _ = file.Close(); _ = os.Remove(lockPath) }()
	return operation()
}
